http/src/main/scala/org/http4s/blaze/http/http2/Http2FrameDecoder.scala

```diff
@@ -0,0 +1,310 @@
+    */
+  def onExtensionFrame(code: Byte, streamId: Int, flags: Byte, buffer: ByteBuffer): Http2Result =
+    Continue
+
+  //////////////// Decoding algorithms ///////////////////////////////////////////////////////////
+
+  //////////// DATA ///////////////
+  // https://tools.ietf.org/html/rfc7540#section-6.1
+  private def decodeDataFrame(buffer: ByteBuffer, streamId: Int, flags: Byte): Http2Result = {
+    if (streamId == 0) {
+      /*
+       DATA frames MUST be associated with a stream.  If a DATA frame is
+       received whose stream identifier field is 0x0, the recipient MUST
+       respond with a connection error (Section 5.4.1) of type
+       PROTOCOL_ERROR.
+       */
+      return Error(PROTOCOL_ERROR.goaway("Data frame with streamID 0x0"))
+    }
+
+    // "The entire DATA frame payload is included in flow control,
+    // including the Pad Length and Padding fields if present."
+    val flowBytes = buffer.remaining
+
+    if (Flags.PADDED(flags)) {
+      val padding = buffer.get() & 0xff
+      val r = limitPadding(padding, buffer)
+      if (!r.success) {
+        return r
+      }
+    }
+
+    listener.onDataFrame(streamId, Flags.END_STREAM(flags), buffer.slice(), flowBytes)
+  }
+
+  //////////// HEADERS ///////////////
+  private def decodeHeaderFrame(buffer: ByteBuffer, streamId: Int, flags: Byte): Http2Result = {
+    if (streamId == 0) {
+      return Error(PROTOCOL_ERROR.goaway("Headers frame with streamID 0x0"))
+    }
+
+    val padding = if (!Flags.PADDED(flags)) 0 else { buffer.get() & 0xff }
+
+    val priority =
+      if (!Flags.PRIORITY(flags)) Priority.NoPriority
+      else getPriority(buffer)
+
+    val r = limitPadding(padding, buffer)
+    if (!r.success) {
+      return r
+    }
+
+    priority match {
+      case Priority.Dependant(dep, _, _) if dep == streamId =>
+        Error(PROTOCOL_ERROR.goaway(s"Header stream depends on itself (id $streamId)"))
+
+      case _ =>
+      listener.onHeadersFrame(
+        streamId, priority, Flags.END_HEADERS(flags),
+        Flags.END_STREAM(flags), buffer.slice())
+    }
+  }
+
+
+  //////////// PRIORITY ///////////////
+  private def decodePriorityFrame(buffer: ByteBuffer, streamId: Int, flags: Byte): Http2Result = {
+    if (streamId == 0) {
+      return Error(PROTOCOL_ERROR.goaway("Priority frame with streamID 0x0"))
+    }
+
+    if (buffer.remaining != 5) {    // Make sure the frame has the right amount of data
+      val msg = "Invalid PRIORITY frame size, required 5, received" + buffer.remaining
+      return Error(FRAME_SIZE_ERROR.rst(streamId, msg))
+    }
+
+    val priority = getPriority(buffer)
+    if (priority.dependentStreamId == streamId) {
+      return Error(PROTOCOL_ERROR.rst(streamId, "Priority frame depends on itself"))
+    }
+
+    listener.onPriorityFrame(streamId, priority)
+  }
+
+  private def getPriority(buffer: ByteBuffer): Priority.Dependant = {
+    val rawInt = buffer.getInt()
+    val priority = (buffer.get() & 0xff) + 1
```

Comment:
A plus one after a mask looks weird to me.  What is this?

Reply:
Spec says the priority is the integer value + 1.